src/wat-writer.cc

```diff
@@ -807,9 +871,46 @@ void WatWriter::PushExpr(const Expr* expr,
     expr_tree_stack_.emplace_back(expr);
     FlushExprTreeStack();
   }
+  if (s_trace) {
+    fprintf(stderr, "-> PushExpr(%s, %" PRIindex ", %" PRIindex")\n",
+            GetExprTypeName(*expr), operand_count, result_count);
+  }
+}
+
+void WatWriter::PushCatch(const Catch* catch_) {
+  if (s_trace)
+    fprintf(stderr, "-> PushCatch()\n");
+  ExprTree tree(catch_);
+  expr_tree_stack_.push_back(std::move(tree));
+  FlushExprTreeStack();
+  if (s_trace)
+    fprintf(stderr, "<- PushCatch()\n");
+}
+
+void WatWriter::FlushCatchTree(const ExprTree& expr_tree) {
+  if (s_trace)
+    fprintf(stderr, "-> FlushCatchTree(%s)\n", expr_tree.describe().c_str());
+  assert(expr_tree.IsCatch());
+  WritePuts("(", NextChar::None);
+  if (expr_tree.catch_->IsCatchAll()) {
+    WritePutsNewline(Opcode::CatchAll_Opcode.GetName());
+  } else {
+    WritePutsSpace(Opcode::Catch_Opcode.GetName());
+    WriteVar(&expr_tree.catch_->var, NextChar::Space);
```

Comment:
Probably should write a newline after this, since it will be followed by an expr list.